src/RestControllers/FHIR/FhirExportRestController.php

```diff
@@ -0,0 +1,574 @@
+                . " occurred during export", ['exception' => $exception->getMessage(),
+                    'trace' => $exception->getTraceAsString(), 'job' => $job->getUuidString(), 'resource' => $resource]);
+            } finally {
+                $this->logger->debug("FhirExportRestController->processResourceExportForJob() closing resource", [
+                    'resource' => $resource, 'recordsExported' => $exportWriter->getRecordsWritten()
+                ]);
+                $exportWriter->close();
+            }
+
+            if (!empty($output)) {
+                $outputResult[] = $output;
+            } else {
+                $errorResult[] = $error;
+            }
+        }
+
+        $processedJob->setOutput(json_encode($outputResult));
+        $processedJob->setErrors(json_encode($errorResult));
+        $processedJob->setStatus(ExportJob::STATUS_COMPLETED);
+
+        return $this->fhirExportJobService->updateJob($processedJob);
+    }
+
+    private function createErrorResultForOutcomeOperation(ExportJob $job, FHIROperationOutcome $errorOutcome)
+    {
+        $jobUuidString = $job->getUuidString();
+        $resource = $errorOutcome->get_fhirElementName();
+        $fileName = $resource  . "-" . $jobUuidString . "-" . $errorOutcome->getId() . ".ndjson";
+        $data = json_encode($errorOutcome);
+        $document = $this->createExportJobFile($job, $fileName, $data);
+        return $this->getResultForResourceDocument($resource, $document);
+    }
+
+    private function createOutputResultForData(ExportJob $job, $resource, &$data)
+    {
+        $jobUuidString = $job->getUuidString();
+        $fileName = $resource . "-" . $jobUuidString . ".ndjson";
+        $document = $this->createExportJobFile($job, $fileName, $data);
+        return $this->getResultForResourceDocument($resource, $document);
+    }
+
+    private function getResultForResourceDocument($resource, \Document $document)
+    {
+        return [
+            'url' => $this->request->getApiBaseFullUrl() . '/fhir/Document/' . $document->get_id() . '/Binary'
+            ,"type" => $resource
+        ];
+    }
+
+    private function createExportJobFile(ExportJob $job, $fileName, &$data): \Document
+    {
+        $document = new \Document();
+        $folder = self::FHIR_DOCUMENT_FOLDER;
+        $categoryId = sqlQuery('Select `id` FROM categories WHERE name=?', [self::FHIR_DOCUMENT_CATEGORY]);
+        if ($categoryId === false) {
+            throw new ExportException("document category id does not exist in system");
+        }
+
+        $mimeType = "application/fhir+ndjson";
+        $higherLevelPath = "";
+        $pathDepth = 1;
+        $owner = 0;  // userID
+        $tmpFile = null;
+
+        $expirationDate = $job->getStartTime()
+            ->add(new \DateInterval(self::MAX_DOCUMENT_ACCESS_TIME))
+            ->format("Y-m-d H:i:s");
+
+        // set the foreign key so we can track documents connected to a specific export
```

Comment:
is the foreign key being set to a patient id or something else? If something else, what are some examples?

Reply:
It's set to the ExportJob's id.  Look at the DeleteRequest in the FhirExportRestController as I need to grab all of the documents for the given ExportJob and mark them all as deleted.

Reply:
The problem using the foreign_id is that this is basically the pid (aka patient_id) mapping row in the documents table, and codebase and custom code will use this when grabbing documents for a patient. So, can't use it to also map something else (ie. or a patient may inadvertently get a document they should not if the mappings clash). Rec considering making another sql column in documents for this mapping which can include as a parameter in the createDocument call. And would make it a KEY/INDEX.

Reply:
I wish foreign_id was named patient_id as that is really bizarre naming convention as I assumed foreign_id could point to any other record type in the system.  I guess we can't change it without updating a ton of code that relies on it.  

I will add a `foreign_reference_id` and a `reference_table` column so we know which table we are pointing at.

Reply:
This should now be fixed.  If you can look over my changes and let me know if anything should be done there, that'd be great.